### DISCUSSION TITLE
Default async handler? Future oder callback?

### DISCUSSION THREAD

**Discussion by lukasberbuer** [open62541pp/open62541pp#256] [Polls]
What should be the default handler for async operations?
The current model is described here: https://github.com/open62541pp/open62541pp/blob/master/doc/async_model.md

What are your opinions @chandryan, @lms-hfr, @pemartin2, @duca, ...?

**Comment by pemartin2** [comment]
I prefer the future approach for read operations. However, the way it's currently implemented it removes the synchrony between read and write interface as write requires a callback use.
Not sure if there is some `std::future` way that would make sense, too. Maybe `std::future<void>` or `std::future<ReturnStruct>`.

**Reply by chandryan** [reply]
I prefer future as default token as well, for all operations :) 
@pemartin2 which function do you refer to for the write interface? There is `opcua::writeDataValueAsync(...)`, for which I suppose it should return an `std::future<Result<void>>` per default.

**Reply by lukasberbuer** [reply]
I just realized, that the default completion token was not defined for those high-level attribute write functions. This might have caused this issue. Fixed in https://github.com/open62541pp/open62541pp/pull/262.
